
<p>Data is collected in the field by taking <strong>samples</strong> of the pathogens from infected patients, recording the <strong>dates</strong> on which these sample are taken, <strong>sequencing</strong> their genomes, and preparing aligned SNP data. These data form the basis for the analysis.</p> 
<p>All four methods use exactly the same <strong>.fasta</strong> files to input the following for each patient (also known as a case or a host):
<ul>
<li>Unique identifier</li>
<li>Sample date</li>
<li>Aligned genetic sequence</li>
</ul>
</p>
<p>
They implement <strong>MCMC</strong> simulations (whose details vary considerably), and output the following information either in log files or via an R library interface:
<ul>
<li>Simulation number</li>
<li>Transmission tree posterior (see 'Discussion' page)</li>
<li>Information on transmission events: who infected who</li>
<li>Timed phylogenetic trees</li> 
</ul>
The phylogenetic trees, transmission trees and transmission networks, corresponding to sampled simulation results, can be seen on the 'Results' page.
</p>
<p>

</p>  

<p></p> 

<p></p>



