<p>Infectious diseases have a global impact with severe human and economic cost. Understanding how and when diseases are transmitted from person to person is vital to identifying methods to control outbreaks.</p>

<p>When a disease outbreak occurs, various questions naturally arise: Who caught it first? Who caught it from whom?  When and where? Traditional epidemiological methods used to answer these questions are labour intensive, expensive and prone to error. Now that genetic sequencing has become relatively cheap, it is possible to sequence the whole genome of the disease agent – be it bacterial or viral – by sampling from each patient. We can then use this timed sequence data to infer both a phylogenetic tree for the disease and the transmission network between patients.</p>  

<p>Various methods have been developed to do this and have been implemented as publicly available algorithms. I compare four of them against each other using two different data sets: one is a small cluster of 13 Tuberculosis (TB) cases in Norway, the other a TB outbreak in Hamburg with 86 known cases.</p> 

<p>The number of possible phylogenetic and transmission trees is truly enormous, even for a small number of patient cases. Nevertheless, using a recently developed measure of distances between trees, I demonstrate that it is possible to meaningfully compare the range of trees that these algorithms produce. This will facilitate their development and testing in future. The ultimate goal is to be able to quickly analyse outbreaks as they occur, enabling rapid interventions to be made in the field to bring them under control.</p>



